### PR TITLE
Feature/temporal query support

### PR DESCRIPTION
**Pull request by rob-blackbourn**
I've added support for temporal table querying. I built this on top of my previous pull request BTW :(

I'm not sure how close this is to how you would like it done.

There are a bunch of examples in the README/tutorial near the bottom. The approach I've taken is to add the criteria to the `Table`. Here's an example:

```python
t = Table("abc")
q = Query.from_(t.for_(SYSTEM_TIME.as_of('2020-01-01'))).select("*")
```

This produces:

```sql
'SELECT * FROM "abc" FOR SYSTEM_TIME AS OF '2020-01-01'
```

There is support for all the valid operations including `UPDATE` and `DELETE`.

A side affect of applying the criteria to the `Table` is that joins are also supported.

Rob

### CODE REVIEW

File: pypika/__init__.py [hunk 1]
```diff
@@ -95,3 +96,6 @@
 __author__ = "Timothy Heys"
 __email__ = "theys@kayak.com"
 __version__ = "0.39.1"
+
+NULL = NullValue()
+SYSTEM_TIME = SystemTimeValue()
```
Comment (mikeengland) [inline pypika/__init__.py:101]: There is a missing newline here. We still need to set up Black formatting at some point.

File: pypika/tests/test_selects.py [hunk 2]
```diff
@@ -302,6 +303,73 @@ def test_table_select_alias_with_offset_and_limit(self):
...
+            q = Query.from_(
+                t.for_(SYSTEM_TIME.from_to('2020-01-01', '2020-02-01'))
+            ).select("*")
+
+            self.assertEqual(
+                'SELECT * FROM "abc" FOR SYSTEM_TIME FROM \'2020-01-01\' TO \'2020-02-01\'',
+                str(q)
+            )
+
+        with self.subTest("with as of ALL"):
```
Comment (mikeengland) [inline pypika/tests/test_selects.py:337]: Minor: Should just be "with ALL"

File: pypika/tests/test_selects.py [hunk 3]
```diff
@@ -302,6 +303,73 @@ def test_table_select_alias_with_offset_and_limit(self):
...
+            q = Query.from_(
+                t.for_(t.valid_period.from_to('2020-01-01', '2020-02-01'))
+            ).select("*")
+
+            self.assertEqual(
+                'SELECT * FROM "abc" FOR "valid_period" FROM \'2020-01-01\' TO \'2020-02-01\'',
+                str(q)
+            )
+
+        with self.subTest("with period as of ALL"):
```
Comment (mikeengland) [inline pypika/tests/test_selects.py:365]: Same as above

**Comment by mikeengland** [comment]
Looks like you have a merge conflict in the PR

File: pypika/tests/test_selects.py [hunk 4]
```diff
@@ -302,6 +303,73 @@ def test_table_select_alias_with_offset_and_limit(self):
               Query.from_("abc").select("foo")[10:10],
         )
 
+    def test_temporal_select(self):
+        t = Table("abc")
```
Comment (mikeengland) [inline pypika/tests/test_selects.py:307]: Would you also mind adding a test for a use case with a join?

**Comment by rob-blackbourn** [comment]
Hi @mikeengland 

I think I've fixed all the things you spotted.

It's worth checking the example of joins I put in the tutorial, as it describes a potential gotcha regarding joins that you might not be happy with, although I think it makes sense.

**Comment by rob-blackbourn** [comment]
I'm a bit rubbish at git. How do I resolve this conflict?

**Comment by mikeengland** [comment]
> I'm a bit rubbish at git. How do I resolve this conflict?

I'd use PyCharm's merge conflict feature by doing  `git pull --rebase upstream master` and then resolving the conflicts inside of PyCharm. It seems you can now also resolve the conflicts inside of Github inside of this PR, but I haven't used that before!

**Comment by x8lucas8x** [comment]
> I'm a bit rubbish at git. How do I resolve this conflict?

If you pull again from master, inside your feature branch, you will see the conflicts locally. So you can just fix the conflicts and update this PR. The conflicts seem pretty basic and were caused by a previous black formatting, which your PR doesn't have applied at the time.

**Comment by rob-blackbourn** [comment]
I used the GitHub GUI like a child :)

**Comment by rob-blackbourn** [comment]
I've formatted with black

**Comment by rob-blackbourn** [comment]
Let me know if there's anything more you need me to do with this.

File: pypika/__init__.py [hunk 5]
```diff
@@ -95,7 +95,7 @@
 
 __author__ = "Timothy Heys"
 __email__ = "theys@kayak.com"
-__version__ = "0.40.0"
+__version__ = "0.41.0"
```
Comment (mikeengland) [inline pypika/__init__.py:98]: Hi @rob-blackbourn, would you mind reverting this as we use the bumpversion tool to manage the versioning, which also updates setup.cfg.

Comment (rob-blackbourn) [reply pypika/__init__.py]: @mikeengland done

**Comment by mikeengland** [comment]
Thanks!

**Pull request merged**